### PR TITLE
[SPARK-27008][SQL] Support java.time.LocalDate as an external type of DateType

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to add new Catalyst type converter for `DateType`. It should be able to convert `java.time.LocalDate` to/from `DateType`.

Main motivations for the changes:
- Smoothly support Java 8 time API
- Avoid inconsistency of calendars used inside of Spark 3.0 (Proleptic Gregorian calendar) and `java.sql.Date` (hybrid calendar - Julian + Gregorian).
- Make conversion independent from current system timezone.

By default, Spark converts values of `DateType` to `java.sql.Date` instances but the SQL config `spark.sql.datetime.java8API.enabled` can change the behavior. If it is set to `true`, Spark uses `java.time.LocalDate` as external type for `DateType`.

## How was this patch tested?

Added new testes to `CatalystTypeConvertersSuite` to check conversion of `DateType` to/from `java.time.LocalDate`, `JavaUDFSuite`/ `UDFSuite` to test usage of `LocalDate` type in Scala/Java UDFs.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
Just FYI: in #23908 I'm proposing refactor of `serializerFor` so either this or #23908 may need to apply other's chance once other one gets merged first.

**Comment by MaxGekk** [comment]
@HeartSaVioR I see, I will rebase.

**Comment by HeartSaVioR** [comment]
Yeah if your patch gets in first I'll apply the change too. Thanks!

**Comment by MaxGekk** [comment]
@cloud-fan @HeartSaVioR Please, review the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1688,6 +1688,12 @@ object SQLConf {
     .stringConf
     .checkValues(Set("Timestamp", "Instant"))
     .createWithDefault("Timestamp")
+
+  val DATE_EXTERNAL_TYPE = buildConf("spark.sql.catalyst.dateType")
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: shall we combine the configs? e.g. `spark.sql.datetime.java8API.enabled`

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: I combined them.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala [hunk 2]
```diff
@@ -393,6 +397,14 @@ object JavaTypeInference {
             inputObject :: Nil,
             returnNullable = false)
 
+        case c if c == classOf[java.sql.Date] && SQLConf.get.datetimeJava8ApiEnabled =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: it's for dataset, so we shouldn't check the config here. If users want to collect a `Dataset[LocalDate]`, Spark must return `LocalDate` no matter what the config value is.

So here we should do
```
case c if c == classOf[LocalDate] =>
  StaticInvoke...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 3]
```diff
@@ -46,7 +46,9 @@ import org.apache.spark.unsafe.types.UTF8String
  *   StringType -> String
  *   DecimalType -> java.math.BigDecimal or scala.math.BigDecimal or Decimal
  *
- *   DateType -> java.sql.Date
+ *   DateType -> java.sql.Date when spark.sql.catalyst.dateType is set to Date
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: let's update the config name

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala]: fixed

**Comment by cloud-fan** [comment]
LGTM

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Review by HeartSaVioR — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by HyukjinKwon** [comment]
Nice, +1!

**Pull request closed without merging**